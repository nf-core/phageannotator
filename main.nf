#!/usr/bin/env nextflow
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    nf-core/phageannotator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Github : https://github.com/nf-core/phageannotator
    Website: https://nf-co.re/phageannotator
    Slack  : https://nfcore.slack.com/channels/phageannotator
----------------------------------------------------------------------------------------
*/

nextflow.enable.dsl = 2

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT FUNCTIONS / MODULES / SUBWORKFLOWS / WORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//
// PLUGIN: Installed from nextflow plugins
//
include { paramsSummaryMap                      } from 'plugin/nf-validation'

//
// MODULE
//
include { MULTIQC                               } from './modules/nf-core/multiqc/main'
include { FASTQC                                } from './modules/nf-core/fastqc/main'

//
// SUBWORKFLOW
//
include { PIPELINE_INITIALISATION       } from './subworkflows/local/utils_nfcore_phageannotator_pipeline'
include { paramsSummaryMultiqc          } from './subworkflows/nf-core/utils_nfcore_pipeline'
include { softwareVersionsToYAML        } from './subworkflows/nf-core/utils_nfcore_pipeline'
include { methodsDescriptionText        } from './subworkflows/local/utils_nfcore_phageannotator_pipeline'
include { PIPELINE_COMPLETION           } from './subworkflows/local/utils_nfcore_phageannotator_pipeline'

//
// WORKFLOW
//
include { PHAGEANNOTATOR            } from './workflows/phageannotator/main'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    NAMED WORKFLOW FOR PIPELINE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//
// WORKFLOW: Run main analysis pipeline depending on type of input
//
workflow NFCORE_PHAGEANNOTATOR {

    take:
    samplesheet // channel: samplesheet read in from --input

    main:

    ch_fastq_gz = samplesheet.map { meta, fastq, fasta -> return [ meta, fastq ] }
    ch_fasta_gz = samplesheet.map { meta, fastq, fasta -> return [ meta, fasta ] }

    //
    // WORKFLOW: Run pipeline
    //
    PHAGEANNOTATOR (
        ch_fastq_gz,
        ch_fasta_gz
    )

    emit:
    versions = PHAGEANNOTATOR.out.versions

}
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow {

    main:
    ch_versions         = Channel.empty()
    ch_multiqc_files    = Channel.empty()

    //
    // SUBWORKFLOW: Run initialisation tasks
    //
    PIPELINE_INITIALISATION (
        params.version,
        params.help,
        params.validate_params,
        params.monochrome_logs,
        args,
        params.outdir,
        params.input
    )

    //
    // MODULE: Run FastQC
    //
    FASTQC (
        PIPELINE_INITIALISATION.out.samplesheet.map { meta, fastq, fasta -> return [ meta, fastq ] }
    )
    ch_multiqc_files = ch_multiqc_files.mix(FASTQC.out.zip.collect{it[1]})
    ch_versions = ch_versions.mix(FASTQC.out.versions.first())

    //
    // WORKFLOW: Run main workflow
    //
    NFCORE_PHAGEANNOTATOR (
        PIPELINE_INITIALISATION.out.samplesheet
    )

    //
    // Collate and save software versions
    //
    softwareVersionsToYAML(NFCORE_PHAGEANNOTATOR.out.versions)
        .collectFile(storeDir: "${params.outdir}/pipeline_info", name: 'nf_core_pipeline_software_mqc_versions.yml', sort: true, newLine: true)
        .set { ch_collated_versions }

    //
    // MODULE: MultiQC
    //
    ch_multiqc_config                     = Channel.fromPath("$projectDir/assets/multiqc_config.yml", checkIfExists: true)
    ch_multiqc_custom_config              = params.multiqc_config ? Channel.fromPath(params.multiqc_config, checkIfExists: true) : Channel.empty()
    ch_multiqc_logo                       = params.multiqc_logo ? Channel.fromPath(params.multiqc_logo, checkIfExists: true) : Channel.empty()
    summary_params                        = paramsSummaryMap(workflow, parameters_schema: "nextflow_schema.json")
    ch_workflow_summary                   = Channel.value(paramsSummaryMultiqc(summary_params))
    ch_multiqc_custom_methods_description = params.multiqc_methods_description ? file(params.multiqc_methods_description, checkIfExists: true) : file("$projectDir/assets/methods_description_template.yml", checkIfExists: true)
    ch_methods_description                = Channel.value(methodsDescriptionText(ch_multiqc_custom_methods_description))
    ch_multiqc_files                      = ch_multiqc_files.mix(ch_workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))
    ch_multiqc_files                      = ch_multiqc_files.mix(ch_collated_versions)
    ch_multiqc_files                      = ch_multiqc_files.mix(ch_methods_description.collectFile(name: 'methods_description_mqc.yaml', sort: false))

    MULTIQC (
        ch_multiqc_files.collect(),
        ch_multiqc_config.toList(),
        ch_multiqc_custom_config.toList(),
        ch_multiqc_logo.toList()
    )

    ch_multiqc_report  = MULTIQC.out.report.toList() // channel: /path/to/multiqc_report.html

    //
    // SUBWORKFLOW: Run completion tasks
    //
    PIPELINE_COMPLETION (
        params.email,
        params.email_on_fail,
        params.plaintext_email,
        params.outdir,
        params.monochrome_logs,
        params.hook_url,
        ch_multiqc_report
    )
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
