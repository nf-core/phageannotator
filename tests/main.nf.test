nextflow_pipeline {

    name "Test pipeline: NFCORE_PHAGEANNOTATOR"
    script "main.nf"
    tag "pipeline"
    tag "nfcore_phageannotator"
    tag "nfcore_phageannotator_default"

    test("Parameters: default") {
        when {
            params {
                outdir                      = "$outputDir"
                assembly_min_length         = 3000
                mash_screen_min_score       = 1
                mash_screen_winner_take_all = true
                genomad_min_score           = 0.7
                genomad_max_fdr             = 1
                genomad_disable_nn          = true
                genomad_sensitivity         = 0.1
                checkv_minimal_db           = true
                checkv_min_length           = 3000
                iphop_partial_test          = true
                max_cpus                    = 4
                max_memory                  = '15.GB'
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    path("${params.outdir}/AssemblyFilter"),
                    path("${params.outdir}/VirusClassification"),
                    path("${params.outdir}/VirusQuality"),
                    path("${params.outdir}/GenomeClustering"),
                    path("${params.outdir}/VirusAbundance/bowtie2/align/test1.bowtie2.log"),
                    path("${params.outdir}/VirusAbundance/bowtie2/build"),
                    file("${params.outdir}/pipeline_info/nf_core_pipeline_software_mqc_versions.yml").readLines()[0..41]
                    ).match() },
                { assert path("${outputDir}/VirusAbundance/coverm/contig/all_samples_alignment_results.tsv").readLines().size() == 3 },
                { assert new File("$outputDir/ReadQuality/fastqc/test1_1_fastqc.html").exists() },
                { assert new File("$outputDir/ReadQuality/fastqc/test1_2_fastqc.html").exists() },
                { assert new File("$outputDir/VirusAbundance/bowtie2/align/test1.bam").exists() },
                { assert new File("$outputDir/multiqc/multiqc_report.html").exists() }
            )
        }
    }
}
