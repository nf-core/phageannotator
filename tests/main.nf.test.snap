{
    "Parameters: default": {
        "content": [
            [
                [
                    [
                        "test1.fasta.gz:md5,e47bee1d12b4f3ecc15357339a300685",
                        "test2.fasta.gz:md5,e47bee1d12b4f3ecc15357339a300685"
                    ]
                ]
            ],
            [
                [
                    [
                        [
                            "test1_virus.fna.gz:md5,3b80368a9cd0add3a1525832106f7df1",
                            "test1_virus_summary.tsv:md5,7e148b000c445c6569e17a0f4556e0f7"
                        ],
                        [
                            "test2_virus.fna.gz:md5,3b80368a9cd0add3a1525832106f7df1",
                            "test2_virus_summary.tsv:md5,7e148b000c445c6569e17a0f4556e0f7"
                        ]
                    ]
                ]
            ],
            [
                [
                    [
                        [
                            "proviruses.fna.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "quality_summary.tsv:md5,67607e20fe485e6e718c956b88d032d7",
                            "viruses.fna.gz:md5,62101302173ac5f830ee4b3ab7435e74"
                        ],
                        [
                            "proviruses.fna.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "quality_summary.tsv:md5,67607e20fe485e6e718c956b88d032d7",
                            "viruses.fna.gz:md5,62101302173ac5f830ee4b3ab7435e74"
                        ]
                    ]
                ],
                [
                    "test1.filtered.fna.gz:md5,89a5812446b42a600dca720e3e0f3eb0",
                    "test2.filtered.fna.gz:md5,89a5812446b42a600dca720e3e0f3eb0"
                ]
            ],
            [
                "all_samples_anicluster_representatives.fna.gz:md5,89a5812446b42a600dca720e3e0f3eb0",
                [
                    "all_samples_ani.tsv:md5,9a9bd8d256c12ee4177c52a030847b7f"
                ],
                [
                    "all_samples_clusters.tsv:md5,0876bb9b6967f272115a1197fe713546"
                ]
            ],
            "test1.bowtie2.log:md5,f161a0eae2eec7fcbfee0d7675d6ef0d",
            [
                [
                    "all_samples_anicluster_representatives.fna.1.bt2:md5,48073ebcfcd5af5002e65d908ac48cb6",
                    "all_samples_anicluster_representatives.fna.2.bt2:md5,eeab4846bdfb11025cce1f0d89b0ab19",
                    "all_samples_anicluster_representatives.fna.3.bt2:md5,7dd079253884f7d2a1cb7b845b6e2bcc",
                    "all_samples_anicluster_representatives.fna.4.bt2:md5,482e19f50cd47fdde58bef83eacd0a3d",
                    "all_samples_anicluster_representatives.fna.rev.1.bt2:md5,a03dd70d091923ff392e3419cc6f96c6",
                    "all_samples_anicluster_representatives.fna.rev.2.bt2:md5,3814fd4a34f1cf9f33bb684cd8fab198"
                ]
            ],
            [
                "ANICLUSTER_ANICALC:",
                "  python: 3.9.12",
                "  biopython: 1.78",
                "  numpy: 1.22.3",
                "ANICLUSTER_ANICLUST:",
                "  python: 3.9.12",
                "ANICLUSTER_EXTRACTREPS:",
                "  python: 3.9.12",
                "  biopython: 1.78",
                "BLAST_MAKEBLASTDB:",
                "  blast: 2.15.0+",
                "BOWTIE2_ALIGN:",
                "  bowtie2: 2.5.2",
                "  samtools: 1.18",
                "  pigz: 2.6",
                "BOWTIE2_BUILD:",
                "  bowtie2: 2.5.2",
                "COVERM_CONTIG:",
                "  coverm: 0.6.1)",
                "GENOMAD_DOWNLOAD:",
                "  genomad: 1.7.4",
                "GUNZIP_CLUSTER_REPS:",
                "  gunzip: 1.1",
                "QUALITYFILTERVIRUSES:",
                "  python: 3.9.12",
                "  biopython: 1.78",
                "  pandas: 1.3.5",
                "SAMTOOLS_FLAGSTAT:",
                "  samtools: 1.19.2",
                "SAMTOOLS_IDXSTATS:",
                "  samtools: 1.19.2",
                "SAMTOOLS_INDEX:",
                "  samtools: 1.19.2",
                "SAMTOOLS_SORT:",
                "  samtools: 1.19.2",
                "SAMTOOLS_STATS:",
                "  samtools: 1.19.2",
                "SEQKIT_SEQ:",
                "  seqkit: 2.2.0",
                "UNTAR:",
                "  untar: 1.3",
                "Workflow:"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-17T15:10:55.472746913"
    }
}